### PR TITLE
[WIP] Eddy Motion Correction (EMC)

### PR DESCRIPTION
**Pull request by dPys**
*Per @arokem 's request to PR the WIP, even though tests + docs are not yet included.
*After an initial glance from folks, plan is to close this WIP PR and eventually separate it into several PR's, perhaps per workflow (and associated interfaces/funcs) @oesteban ?
*About 50% of the code is adapted from fmriprep utilities and qsiprep's implementation of 3dShore, & around 50% is new.
*Supports 3dShore EMC for multishell data, but now implements it exclusively in Dipy as opposed to BrainSuite.
*More importantly, can implement Tensor and SparseFascicle model for LOO prediction on single-shell data! 
*Registration routines currently use ANTs, but could probably be adapted to use to Dipy instead. In that case, they would most likely require careful retuning.
*Pictures to come, but initial visual checks seem to indicate at least decent quality performance (perhaps even reasonably good performance) as a starting benchmark.

### CODE REVIEW

File: dmriprep/config/emc_coarse_Affine.json [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+    "write_composite_transform": false,
+    "use_histogram_matching": [ false, false ],
+    "use_estimate_learning_rate_once": [ true, true ],
+    "transforms": [ "Rigid", "Affine" ],
+    "number_of_iterations": [ [ 100, 100 ], [ 100 ] ],
+    "output_warped_image": true,
+    "transform_parameters": [ [ 0.2 ], [ 0.15 ] ],
+    "convergence_threshold": [ 1e-06, 1e-06 ],
+    "convergence_window_size": [ 20, 20 ],
+    "metric": [ "Mattes", "Mattes" ],
```
Comment (oesteban) [inline dmriprep/config/emc_coarse_Affine.json]: Is this choice based on speed?

Wouldn't it make more sense some cross-correlation based measure? MI is just going to exacerbate the problems you mentioned about the different content of each direction (pinging @arokem and @mattcieslak).

File: dmriprep/interfaces/bids.py [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+                desc=desc,
+                suffix=suffix,
+                i=i,
+                dtype=dtype,
+                ext=ext)
+            if isdefined(self.inputs.extra_values):
+                out_file = out_file.format(extra_value=self.inputs.extra_values[i])
+            self._results['out_file'].append(out_file)
+            self._results['compression'].append(_copy_any(fname, out_file))
+        return runtime
```
Comment (oesteban) [inline dmriprep/interfaces/bids.py]: We already have a DerivativesDataSink:

https://github.com/nipreps/dmriprep/blob/0a609b4c91c97b6b2bb0fad0f35ce2e7d15f44b3/dmriprep/interfaces/__init__.py#L13-L16

Unless there is some new functionality here, this can go away.

File: dmriprep/interfaces/images.py [hunk 3]
```diff
@@ -45,31 +51,13 @@ def _run_interface(self, runtime):
...
-
-
 class _RescaleB0InputSpec(BaseInterfaceInputSpec):
     in_file = File(exists=True, mandatory=True, desc='b0s file')
     mask_file = File(exists=True, mandatory=True, desc='mask file')
 
 
 class _RescaleB0OutputSpec(TraitedSpec):
-    out_ref = File(exists=True, desc='One average b0 file')
+    out_ref = File(exists=True, desc='One median b0 file')
```
Comment (oesteban) [inline dmriprep/interfaces/images.py]: Please avoid changes out of scope. This one should've been picked up during the review of #25 - that said, I think it is better "average" here, as it needs to be informative.

File: dmriprep/interfaces/images.py [hunk 4]
```diff
@@ -103,43 +91,279 @@ def _run_interface(self, runtime):
...
+            if not isdefined(output):
+                _, fname, ext = split_filename(self.inputs.in_file)
+                output = fname + "_" + self.inputs.operation + ext
+            return output
+        return None
+
+    def _list_outputs(self):
+        outputs = self.output_spec().get()
+        outputs['out_file'] = os.path.abspath(self._gen_filename('out_file'))
+        return outputs
```
Comment (oesteban) [inline dmriprep/interfaces/images.py]: Look at that! https://github.com/poldracklab/niworkflows/blob/4e9181fd31f3a1b216b5edcf040bd5533c400e9d/niworkflows/interfaces/ants.py#L11-L77

Please use ImageMath from niworkflows (and PR against niworkflows if some functionality is missing)

File: dmriprep/interfaces/vectors.py [hunk 5]
```diff
@@ -92,8 +99,57 @@ def _run_interface(self, runtime):
...
+        cwd = Path(runtime.cwd).absolute()
+        reor_rasb_file = fname_presuffix(
+            self.inputs.rasb_file, use_ext=False, suffix='_reoriented.tsv',
+            newpath=str(cwd))
+        np.savetxt(str(reor_rasb_file), reor_table,
+                   delimiter='\t', header='\t'.join('RASB'),
+                   fmt=['%.8f'] * 3 + ['%g'])
+
+        self._results['out_rasb'] = reor_rasb_file
+        return runtime
```
Comment (oesteban) [inline dmriprep/interfaces/vectors.py]: Isn't this part of #49 ?

File: dmriprep/utils/bids.py [hunk 6]
```diff
@@ -160,3 +160,50 @@ def validate_input_dir(exec_env, bids_dir, participant_label):
...
+        with open(dst, 'wb') as f_out:
+            if dst_isgz:
+                # Remove FNAME header from gzip (poldracklab/fmriprep#1480)
+                gz_out = gzip.GzipFile('', 'wb', 9, f_out, 0.)
+                copyfileobj(f_in, gz_out)
+                gz_out.close()
+            else:
+                copyfileobj(f_in, f_out)
+
+    return True
```
Comment (oesteban) [inline dmriprep/utils/bids.py]: I think this comes from nipype, why not just import them?

File: dmriprep/utils/images.py [hunk 7]
```diff
@@ -0,0 +1,113 @@
+import numpy as np
```
Comment (oesteban) [inline dmriprep/utils/images.py]: This new file makes a lot of sense, especially if we can come up with unit tests for all these helper functions.

Could you send a separate PR to split the currently existing functions (extract_b0 and rescale_b0) out to this file so they don't show up in the diff?

File: dmriprep/utils/vectors.py [hunk 8]
```diff
@@ -375,3 +375,109 @@ def bvecs2ras(affine, bvecs, norm=True, bvec_norm_epsilon=0.2):
...
+
+
+def reorient_vecs_from_ras_b(rasb_file, affines, b0_threshold=B0_THRESHOLD):
+    """
+    Reorient the vectors from a rasb .tsv file.
+    When correcting for motion, rotation of the diffusion-weighted volumes
+    might cause systematic bias in rotationally invariant measures, such as FA
+    and MD, and also cause characteristic biases in tractography, unless the
+    gradient directions are appropriately reoriented to compensate for this
+    effect [Leemans2009]_.
```
Comment (oesteban) [inline dmriprep/utils/vectors.py]: Undefined reference

File: dmriprep/utils/vectors.py [hunk 9]
```diff
@@ -375,3 +375,109 @@ def bvecs2ras(affine, bvecs, norm=True, bvec_norm_epsilon=0.2):
         rotated_bvecs[~b0s] /= norms_bvecs[~b0s, np.newaxis]
         rotated_bvecs[b0s] = np.zeros(3)
     return rotated_bvecs
+
+
+def reorient_vecs_from_ras_b(rasb_file, affines, b0_threshold=B0_THRESHOLD):
```
Comment (oesteban) [inline dmriprep/utils/vectors.py]: I would try to keep the b0 threshold encoded in the vector representation object, meaning, no need to keep dragging it all around. Please remove.

Comment (oesteban) [reply dmriprep/utils/vectors.py]: Actually, I would incorporate the reorientation into the vectors object. Will comment on this at #49. We definitely should push that one first.

File: dmriprep/utils/vectors.py [hunk 10]
```diff
@@ -375,3 +375,109 @@ def bvecs2ras(affine, bvecs, norm=True, bvec_norm_epsilon=0.2):
...
+        M[1, 0] = aff_mat['AffineTransform_float_3_3'][3]
+        M[1, 1] = aff_mat['AffineTransform_float_3_3'][4]
+        M[1, 2] = aff_mat['AffineTransform_float_3_3'][5]
+        M[2, 0] = aff_mat['AffineTransform_float_3_3'][6]
+        M[2, 1] = aff_mat['AffineTransform_float_3_3'][7]
+        M[2, 2] = aff_mat['AffineTransform_float_3_3'][8]
+        M[3, 3] = 1
+        M[0:3, 3] = aff_mat['fixed'].T
+        affines_ras.append(np.multiply(M, ras_trans))
+        del M
```
Comment (oesteban) [inline dmriprep/utils/vectors.py]: will comment on #49, but one thing is rotating the vectors and the other is reading a transform. This should be done separately.

**Comment by dPys** [comment]
Thank you @oesteban for all of your helpful feedback. I will be out of town next week so won't be able to work a ton on this after tomorrow. But will do my best to keep the dev of EMC moving...

In the meantime, @arokem -- anything that we can think of to speed up the way that sfm predicts signal (i.e. from a given vector coordinate) would be helpful. If it can't be the ElasticNet itself, then we might have to get creative to make it viable with LOO here. Please let me know if there's anything that can be done.

More soon.
@dPys

**Comment by mattcieslak** [comment]
@arokem @dPys one thing that made shoreline reasonably fast with 3dSHORE is that I used the L2 (with LShore and NShore) fit instead of L1 or elasticnet. This reduces the problem to a couple dot products and still resulted in very good signal prediction. Would this work with a sfm?

**Comment by arokem** [comment]
Yes. That's a good idea and my hunch is that it would work well: Ridge SFM!
Not too hard to implement either as the DIPY SFM object can take a solver
as an input.

File: dmriprep/config/emc_coarse_Affine.json (hunk 1 above)
Comment (dPys) [reply dmriprep/config/emc_coarse_Affine.json]: I agree @oesteban -- cross-corr seems like it would be preferable here. Curious what @mattcieslak thinks?

File: dmriprep/interfaces/bids.py (hunk 2 above)
Comment (dPys) [reply dmriprep/interfaces/bids.py]: yep!

File: dmriprep/interfaces/images.py (hunk 3 above)
Comment (dPys) [reply dmriprep/interfaces/images.py]: Sounds good.

File: dmriprep/interfaces/images.py (hunk 4 above)
Comment (dPys) [reply dmriprep/interfaces/images.py]: Awesome

File: dmriprep/interfaces/vectors.py (hunk 5 above)
Comment (dPys) [reply dmriprep/interfaces/vectors.py]: Yes, but with added functionality. All future changes for this interface will be redirected to the #49 PR

File: dmriprep/utils/bids.py (hunk 6 above)
Comment (dPys) [reply dmriprep/utils/bids.py]: Didn't see these functions in nipype, but maybe I missed them somewhere?

File: dmriprep/utils/images.py (hunk 7 above)
Comment (dPys) [reply dmriprep/utils/images.py]: Yes.

File: dmriprep/utils/vectors.py (hunk 8 above)
Comment (dPys) [reply dmriprep/utils/vectors.py]: I can just removed this reference, as b-vector rotation is fairly common practice/knowledge at this point.

File: dmriprep/utils/vectors.py (hunk 9 above)
Comment (dPys) [reply dmriprep/utils/vectors.py]: That makes sense

File: dmriprep/utils/vectors.py (hunk 10 above)
Comment (dPys) [reply dmriprep/utils/vectors.py]: In my latest push, I've separated them out into multiple functions.

File: dmriprep/interfaces/images.py (hunk 4 above)
Comment (dPys) [reply dmriprep/interfaces/images.py]: So the reason for redefining it outside of niworkflows is that we need it to be able to take an explicit secondary argument and file. In other words, as defined here, it is actually quite different. But we could monkey-patch it?

Comment (josephmje) [reply dmriprep/interfaces/images.py]: Yes, we can import the niworkflows version and then add to it similar to the `DerivativesDataSink` and `BIDSDataGrabberOutputSpec` examples here: https://github.com/nipreps/dmriprep/blob/master/dmriprep/interfaces/__init__.py#L13-L20

File: dmriprep/config/emc_coarse_Affine.json (hunk 1 above)
Comment (mattcieslak) [reply dmriprep/config/emc_coarse_Affine.json]: Remember, the fixed (model-estimated) images here will have the same contrast as the moving images so most metrics would be ok here. MI is much faster than CC, so any benefit of CC would be outweighed by its extra computation time. Also, this is only the coarse iteration.

**Comment by dPys** [comment]
@arokem @oesteban @mattcieslak -- The EMC WIP should now run from start to finish. I've also gone ahead and replaced every ANTs reg call with a dipy / pure-python equivalent. These new reg routines are based on @arokem 's api.py file, but have been modified to read and write to files from disk. The routines have also been more or less tuned to the original coarse/precise rigid/affine reg parameters used with the ANTs calls. 

Another note-- sfm is now incorporated with a ridge/L2 solver, which made prediction much faster, among other modifications.

With these further upgrades, I'm seeing a dramatic speedup as well for EMC as a whole. 

To get a sense of EMC's effectiveness at this point, here is a before-and-after:
Screen Shot 2020-02-04 at 11 19 58 AM



Now, to actually PR this beast, I propose the following six separate PR's (@oesteban -- please confirm):
1. Reorient vectors interface (FYI:  now relies on a method that has been integrated into the DiffusionGradientTable class)
2. Registration routines (currently included as a separate module register.py in both utils and interfaces)
3. Signal prediction interface 
4. Miscellaneous image-handling utilities and interfaces (i.e. utils/images.py and interfaces/images.py) 
5. Motion reports interfaces
6. EMC workflows (i.e. emc.py)

File: dmriprep/config/emc_coarse_Affine.json (hunk 1 above)
Comment (oesteban) [reply dmriprep/config/emc_coarse_Affine.json]: Yup, but MI has much flatter gradient - I'm going to make up a number here, but 1 CC iteration is as effective as 10 MI iterations.

**Comment by oesteban** [comment]
Taking up from where @dPys left off:

> 1. Reorient vectors interface (FYI: now relies on a method that has been integrated into the DiffusionGradientTable class)

I think we have pretty much all we need already merged. Will look back at this if something is missing.

> 2. Registration routines (currently included as a separate module register.py in both utils and interfaces)

@arokem is working on this.

> 3. Signal prediction interface

I'll extract the code from this PR and separate on to a new one.

> 4. Miscellaneous image-handling utilities and interfaces (i.e. utils/images.py and interfaces/images.py)

I think this is already done in part of in full by @josephmje. As for 1, will come back to this PR if anything is needed.

> 5. Motion reports interfaces + figures/viz

Let's leave this out for now, this is being worked out on a different repo.

> 6. EMC workflows (i.e. emc.py)

Will extract this after 3 is merged.

**Comment by oesteban** [comment]
Okay, pushed Derek's branch (and his commits) to https://github.com/nipreps/dmriprep/tree/feat/wip/hmc-dpys, just to be sure we don't lose any lines.

**Comment by oesteban** [comment]
Closing for clarity.
